### PR TITLE
FIX-#6294: implement `__iter__` for `array` class that defaults to numpy

### LINKED ISSUE #6294: Implement `__iter__` for `array` class
Iteration over a distributed object works very slowly, it is better to default in numpy, similar to how it is done for a Series.

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6294 
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Review by RehanSD — CHANGES_REQUESTED** [review]
Thank you @anmyachev - this fix looks good, but I think we want to defer the specific computation to the query compiler - that way we support other query compilers that may not want to move data out of distributed storage, or may incur costs doing so. Would you mind making the API Layer for the Array, Series, and DF route to a query_compiler method, and have the query_compiler method just return this? We could have it look something like this maybe?
```python
def __iter__(self, lib="pandas", squeeze=False):
     return self.to_numpy().__iter__() if lib != "pandas" else (self.to_pandas().squeeze() if squeeze else self.to_pandas().__iter__()
```

**Comment by anmyachev** [comment]
> Thank you @anmyachev - this fix looks good, but I think we want to defer the specific computation to the query compiler - that way we support other query compilers that may not want to move data out of distributed storage, or may incur costs doing so. Would you mind making the API Layer for the Array, Series, and DF route to a query_compiler method, and have the query_compiler method just return this? We could have it look something like this maybe?
> 
> ```python
> def __iter__(self, lib="pandas", squeeze=False):
>      return self.to_numpy().__iter__() if lib != "pandas" else (self.to_pandas().squeeze() if squeeze else self.to_pandas()).__iter__()
> ```

ok, will do

**Comment by anmyachev** [comment]
> ```python
> def __iter__(self, lib="pandas", squeeze=False):
>      return self.to_numpy().__iter__() if lib != "pandas" else (self.to_pandas().squeeze() if squeeze else self.to_pandas()).__iter__()
> ```

@RehanSD this will also require `Series._to_pandas` and `array._to_numpy` methods to be moved to the query compiler. `query_compiler.to_pandas` and `query_compiler.to_numpy` are not suitable. I would like to clarify whether you think it is necessary to continue moving methods in such a situation.

File: modin/numpy/arr.py [hunk 1]
```diff
@@ -734,6 +734,16 @@ def __abs__(
...
 
+    def __iter__(self):
+        """
+        Return an iterator of the values.
+
+        Returns
+        -------
+        iterable
+        """
+        return self._to_numpy().__iter__()
```
Comment (anmyachev) [inline modin/numpy/arr.py:745]: @RehanSD I can implement in this way. This is more flexible, and there is no need to transfer the implementation of `_to_numpy` function to the query compiler. What do you think?

```suggestion
        if hasattr(self._query_compiler, "numpy__iter__"):
            return self._query_compiler.numpy__iter__()
        return self._to_numpy().__iter__()
```

Comment (vnlitvinov) [reply modin/numpy/arr.py:745]: I personally do not like the idea of such a dynamic interface... I would rather this `numpy__iter__()` would do this `._to_numpy().__iter__()` thing itself...

That said, why are we even wanting to default to pandas (or numpy) in such a simple case? Maybe just implement the iterator ourselves without even defaulting?

Comment (anmyachev) [reply modin/numpy/arr.py:745]: It probably makes sense to implement an iterator, but right now I don't have the ability to do it.

**Pull request closed without merging**